File: vital/tests/test_camera_rpc.cxx
/*ckwg +29
 * Copyright 2014-2018 by Kitware, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  * Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 *  * Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
 *    to endorse or promote products derived from this software without specific
 *    prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * \file
 * \brief test core camera class
 */

#include <test_eigen.h>
#include <tests/test_gtest.h>

#include <vital/types/camera_rpc.h>
#include <vital/io/camera_io.h>
#include <vital/tests/rpc_reader.h>

#include <iostream>

static double epsilon = 1e-8;
kwiver::vital::path_t g_data_dir;

// ----------------------------------------------------------------------------
int main(int argc, char** argv)
{
  ::testing::InitGoogleTest( &argc, argv );
  GET_ARG(1, g_data_dir);
  return RUN_ALL_TESTS();
}

// ----------------------------------------------------------------------------
class camera_rpc : public ::testing::Test
{
  TEST_ARG(data_dir);

public:
  void SetUp()
  {
    // Actual world values
    static const size_t num_points = 5;
    double act_x[num_points] =
      { -58.58940727826357, -58.589140738420539, -58.588819506933184, -58.58855693683482, -58.58839238727699 };
    double act_y[num_points] =
      { -34.49283455146763, -34.492818509990848, -34.492808611762605, -34.492802905977392, -34.49280925602671 };
    double act_z[num_points] =
      { 20.928231142319902, 21.9573811423199, 27.1871011423199, 19.2657311423199, 26.606641142319901 };

    // Actual projection values
    double act_u[num_points] =
      { 16581.12626986, 16519.24664854, 16449.76676766, 16377.35597454, 16347.72126206 };
    double act_v[num_points] =
      { 15443.08533878, 15451.02512727, 15458.40044985, 15461.20973047, 15462.29884238 };

    for (size_t i = 0; i<num_points; ++i)
    {
      test_points.push_back(
        kwiver::vital::vector_3d( act_x[i], act_y[i], act_z[i] ) );
      test_image_points.push_back(
        kwiver::vital::vector_2d( act_u[i], act_v[i] ) );
    }
  }

  std::vector<kwiver::vital::vector_3d> test_points;
  std::vector<kwiver::vital::vector_2d> test_image_points;
};

// ----------------------------------------------------------------------------
TEST_F(camera_rpc, identity)
{
  kwiver::vital::simple_camera_rpc icam;

  kwiver::vital::vector_3d pt(1.0, 2.0, 10.0);

  auto img_pt = icam.project( pt );

  EXPECT_MATRIX_EQ( img_pt, kwiver::vital::vector_2d(1.0, 2.0) );
}

// ----------------------------------------------------------------------------
TEST_F(camera_rpc, clone)
{
  kwiver::vital::path_t test_rpc_file = data_dir + "/rpc_data.dat";
  auto cam = read_rpc( test_rpc_file );
  auto cam_clone =
    std::dynamic_pointer_cast<kwiver::vital::camera_rpc>( cam.clone() );

  EXPECT_MATRIX_EQ( cam.world_scale(), cam_clone->world_scale() );
  EXPECT_MATRIX_EQ( cam.world_offset(), cam_clone->world_offset() );
  EXPECT_MATRIX_EQ( cam.image_scale(), cam_clone->image_scale() );
  EXPECT_MATRIX_EQ( cam.image_offset(), cam_clone->image_offset() );
  EXPECT_MATRIX_EQ( cam.rpc_coeffs(), cam_clone->rpc_coeffs() );
  EXPECT_EQ( cam.image_width(), cam_clone->image_width() );
  EXPECT_EQ( cam.image_height(), cam_clone->image_height() );
}

// ----------------------------------------------------------------------------
TEST_F(camera_rpc, projection)
{
  kwiver::vital::path_t test_rpc_file = data_dir + "/rpc_data.dat";
  auto cam = read_rpc( test_rpc_file );

  for (size_t i = 0; i < test_points.size(); ++i)
  {
    auto img_pt = cam.project( test_points[i] );

    EXPECT_MATRIX_NEAR( img_pt, test_image_points[i], epsilon );
  }
}

// ----------------------------------------------------------------------------
TEST_F(camera_rpc, back_projection)
{
  kwiver::vital::path_t test_rpc_file = data_dir + "/rpc_data.dat";
  auto cam = read_rpc( test_rpc_file );

  for (size_t i = 0; i < test_points.size(); ++i)
  {
    auto img_pt = cam.project( test_points[i] );
    auto new_pt = cam.back_project( img_pt, test_points[i][2] );

    EXPECT_MATRIX_NEAR( new_pt, test_points[i], epsilon );
  }
}

// ----------------------------------------------------------------------------
TEST_F(camera_rpc, read_missing_image_dimension)
{
  kwiver::vital::path_t test_rpc_file = data_dir + "/rpc_data_missing_image_dimension.dat";
  auto cam = read_rpc( test_rpc_file );
  EXPECT_EQ( cam.image_width(), 0 );
  EXPECT_EQ( cam.image_height(), 0 );
}

File: python/kwiver/vital/tests/test_camera_rpc.py
"""
ckwg +31
Copyright 2020 by Kitware, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.

 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

 * Neither name of Kitware, Inc. nor the names of any contributors may be used
   to endorse or promote products derived from this software without specific
   prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

==============================================================================

Tests for CameraRPC interface class.

"""
import unittest
import nose.tools as nt
import numpy as np
import math

from kwiver.vital.tests.py_helpers import no_call_pure_virtual_method
from kwiver.vital.tests.cpp_helpers import camera_rpc_helpers as crpch
from kwiver.vital.types import CameraRPC as crpc, SimpleCameraRPC as srpc



class TestCameraRPC(unittest.TestCase):
    def test_construct(self):
        crpc()

    def test_pure_virts(self):
        no_call_pure_virtual_method(crpc().project, np.array([1.0, 2.0, 10.0]))
        no_call_pure_virtual_method(crpc().back_project, np.array([1.0, 2.0]), 1.0)
        no_call_pure_virtual_method(crpc().rpc_coeffs)
        no_call_pure_virtual_method(crpc().world_scale)
        no_call_pure_virtual_method(crpc().world_offset)
        no_call_pure_virtual_method(crpc().image_scale)
        no_call_pure_virtual_method(crpc().image_offset)
        no_call_pure_virtual_method(crpc().image_height)
        no_call_pure_virtual_method(crpc().image_width)


class TestSimpleCameraRPC(unittest.TestCase):
    @classmethod
    def setUp(self):
        self.m_world_scale = np.array([0, 1, 0])
        self.m_world_offset = np.array([1, 0, 1])
        self.m_image_scale = np.array([1, 1])
        self.m_image_offset = np.array([0, 2])
        self.m_image_width = 1080
        self.m_image_height = 720
        self.m_coeffs = np.ndarray(shape=(4, 20),dtype=float)
        self.m_srpc = srpc(self.m_world_scale, self.m_world_offset, self.m_image_scale,
                                 self.m_image_offset, self.m_coeffs,
                                 self.m_image_width, self.m_image_height)

    def test_constructors(self):
        srpc()
        srpc(self.m_world_scale, self.m_world_offset, self.m_image_scale,
                                 self.m_image_offset, self.m_coeffs,
                                 self.m_image_width, self.m_image_height)
        srpc(srpc())
        srpc(srpc(self.m_world_scale, self.m_world_offset, self.m_image_scale,
                                 self.m_image_offset, self.m_coeffs,
                                 self.m_image_width, self.m_image_height))

    def test_methods(self):
        cameraRPC = srpc()

        # test setters && getters (getters inherited from camera_rpc)
        cameraRPC.set_rpc_coeffs(self.m_coeffs)
        np.testing.assert_array_equal(self.m_coeffs, cameraRPC.rpc_coeffs())
        cameraRPC.set_world_scale(self.m_world_scale)
        np.testing.assert_array_equal(self.m_world_scale, cameraRPC.world_scale())
        cameraRPC.set_world_offset(self.m_world_offset)
        np.testing.assert_array_equal(self.m_world_offset, cameraRPC.world_offset())
        cameraRPC.set_image_scale(self.m_image_scale)
        np.testing.assert_array_equal(self.m_image_scale, cameraRPC.image_scale())
        cameraRPC.set_image_offset(self.m_image_offset)
        np.testing.assert_array_equal(self.m_image_offset, cameraRPC.image_offset())
        cameraRPC.set_image_width(self.m_image_width)
        np.testing.assert_array_equal(self.m_image_width, cameraRPC.image_width())
        cameraRPC.set_image_height(self.m_image_height)
        np.testing.assert_array_equal(self.m_image_height, cameraRPC.image_height())

        # test project, back_project, jacobian
        proj_2d = srpc().project(np.array([1.0, 2.0, 10.0]))
        np.testing.assert_array_equal(proj_2d, np.array([1.0, 2.0]))
        proj_3d = srpc().back_project([1.0, 2.0], 1.0)
        np.testing.assert_array_equal(np.array([1.0, 2.0 ,1.0]), proj_3d)
        ret_val = np.array([0, 0])
        ret_jac = np.ndarray(shape=(2, 2), dtype=(float), order='F')
        ret_jac.setflags(write=1)
        rpc_coeffs_buffer = [[0.006585953, -1.032582, 0.001740937, 0.03034485, 0.0008819178, -0.000167943, 0.0001519299 -0.00626254, -0.00107337, 9.099077e-06, 2.608985e-06, -2.947004e-05, 2.231277e-05, 4.587831e-06, 4.16379e-06, 0.0003464555, 3.598323e-08, -2.859541e-06, 5.159311e-06, -1.349187e-07],
         [1, 0.0003374458, 0.0008965622, -0.0003730697, -2.666499e-05, -2.711356e-06, 5.454434e-07, 4.485658e-07, 2.534922e-05, -4.546709e-06, 0, -1.056044e-07, -5.626866e-07, 2.243313e-08, -2.108053e-07, 9.199534e-07, 0, -3.887594e-08, -1.437016e-08, 0],
         [0.0002703625, 0.04284488, 1.046869, 0.004713542, -0.0001706129, -1.525177e-07, 1.255623e-05, -0.0005820134,  -0.000710512, -2.510676e-07, 3.179984e-06, 3.120413e-06, 3.19923e-05, 4.194369e-06, 7.475295e-05, 0.0003630791, 0.0001021649, 4.493725e-07, 3.156566e-06, 4.596505e-07],
         [1, 0.0001912806, 0.0005166397, -1.45044e-05, -3.860133e-05, 2.634582e-06, -4.551145e-06, 6.859296e-05, -0.0002410782, 9.753265e-05, -1.456261e-07, 5.310624e-08, -1.913253e-05, 3.18203e-08, 3.870586e-07, -0.000206842, 9.128349e-08, 0, -2.506197e-06, 0]]

        rpc_coeffs = np.ndarray(shape = (4,20), dtype = float)
        for x in range(len(rpc_coeffs_buffer)):
            for y in range(len(rpc_coeffs_buffer[x])):
                rpc_coeffs[x][y] = rpc_coeffs_buffer[x][y]
        world_scale = np.array([0.0928, 0.0708, 501])
        world_offset = np.array([-58.6096, -34.4732, 31])
        image_scale = np.array([21250, 21478])
        image_offset = np.array([21249, 21477])
        image_width = 1000
        image_height = 1000
        s = srpc(world_scale, world_offset, image_scale, image_offset, rpc_coeffs, image_height, image_width)
        ret_jac, ret_val = s.jacobian(np.array([-58.58, -34.49, 20.92]), ret_jac, ret_val)

        self.assertFalse(ret_jac.flags.owndata)
        np.testing.assert_array_almost_equal(ret_jac[1], np.array([ 2.92836539e-002, -2.25968e-001]))

        # test clone
        cameraRPC_2 = self.m_srpc.clone()
        # test for slice on c++ side
        self.assertIsInstance(cameraRPC_2, srpc)
        # test for value copying
        np.testing.assert_array_equal(self.m_coeffs, cameraRPC_2.rpc_coeffs())
        np.testing.assert_array_equal(self.m_world_scale, cameraRPC_2.world_scale())
        np.testing.assert_array_equal(self.m_world_offset, cameraRPC_2.world_offset())
        np.testing.assert_array_equal(self.m_image_scale, cameraRPC_2.image_scale())
        np.testing.assert_array_equal(self.m_image_offset, cameraRPC_2.image_offset())
        np.testing.assert_array_equal(self.m_image_width, cameraRPC_2.image_width())
        np.testing.assert_array_equal(self.m_image_height, cameraRPC_2.image_height())

    def test_call_overrides(self):
        # clone
        clone_ = crpch.call_clone(self.m_srpc)
        self.assertIsInstance(clone_, srpc)
        np.testing.assert_array_equal(self.m_coeffs, clone_.rpc_coeffs())
        np.testing.assert_array_equal(self.m_world_scale, clone_.world_scale())
        np.testing.assert_array_equal(self.m_world_offset, clone_.world_offset())
        np.testing.assert_array_equal(self.m_image_scale, clone_.image_scale())
        np.testing.assert_array_equal(self.m_image_offset, clone_.image_offset())
        np.testing.assert_array_equal(self.m_image_width, clone_.image_width())
        np.testing.assert_array_equal(self.m_image_height, clone_.image_height())

        # other members
        np.testing.assert_array_equal(self.m_coeffs, crpch.call_rpc_coeffs(self.m_srpc))
        np.testing.assert_array_equal(self.m_world_scale, crpch.call_world_scale(self.m_srpc))
        np.testing.assert_array_equal(self.m_world_offset, crpch.call_world_offset(self.m_srpc))
        np.testing.assert_array_equal(self.m_image_width, crpch.call_image_width(self.m_srpc))
        np.testing.assert_array_equal(self.m_image_height, crpch.call_image_height(self.m_srpc))
        np.testing.assert_array_equal(self.m_image_scale, crpch.call_image_scale(self.m_srpc))
        np.testing.assert_array_equal(self.m_image_offset, crpch.call_image_offset(self.m_srpc))
        np.testing.assert_array_equal(np.array([1.0, 2.0]), crpch.call_project(srpc(), [1.0, 2.0, 10.0]))
        np.testing.assert_array_equal(np.array([1.0, 2.0, 1.0]), crpch.call_back_project(srpc(), np.array([1.0, 2.0]), 1.0))

class InheritedRPC(crpc):
    def __init__(self):
        crpc.__init__(self)

    def clone(self):
        return InheritedRPC()

    def rpc_coeffs(self):
        return np.ndarray(shape=(4, 20), dtype=float)

    def world_offset(self):
        return np.array([1, 8, 9])

    def world_scale(self):
        return np.array([8, 9, 10])

    def image_scale(self):
        return np.array([0.5, 2.9])

    def image_offset(self):
        return np.array([1.0, 0.75])

    def image_width(self):
        return 3840

    def image_height(self):
        return 2160

    def project(self, pt):
        return np.array([99, 100])

    def back_project(self, pt, elev):
        return np.array([1, 2, 3])


class TestInheritedRPC(unittest.TestCase):
    def test_construct(self):
        InheritedRPC()

    def test_inheritance(self):
        nt.ok_(issubclass(InheritedRPC, crpc))

    def test_clone(self):
        irpc = InheritedRPC()
        cloned_ = crpch.call_clone(irpc)
        self.assertIsInstance(cloned_, InheritedRPC)
        self.assertEqual(cloned_.image_width(), InheritedRPC().image_width())

    def test_methods(self):
        irpc = InheritedRPC()
        self.assertEqual(crpch.call_rpc_coeffs(irpc).shape, np.ndarray(shape=(4, 20), dtype=float).shape)
        np.testing.assert_array_equal(crpch.call_world_scale(irpc), np.array([8, 9, 10]))
        np.testing.assert_array_equal(crpch.call_world_offset(irpc), np.array([1, 8, 9]))
        np.testing.assert_array_equal(crpch.call_image_scale(irpc), np.array([0.5, 2.9]))
        np.testing.assert_array_equal(crpch.call_image_offset(irpc), np.array([1.0, 0.75]))
        self.assertEqual(crpch.call_image_width(irpc), 3840)
        self.assertEqual(crpch.call_image_height(irpc), 2160)
        np.testing.assert_array_equal(crpch.call_project(irpc, np.array([1, 2, 3])), np.array([99, 100]))
        np.testing.assert_array_equal(crpch.call_back_project(irpc, np.array([0, 0.125]), 1), np.array([1, 2, 3]))
